mypy/stubgen.py

```diff
@@ -740,80 +480,120 @@ def visit_func_def(self, o: FuncDef) -> None:
-            annotation = ""
+            typename = ""
             if annotated_type and not is_self_arg and not is_cls_arg:
                 # Luckily, an argument explicitly annotated with "Any" has
                 # type "UnboundType" and will not match.
                 if not isinstance(get_proper_type(annotated_type), AnyType):
-                    annotation = f": {self.print_annotation(annotated_type)}"
+                    typename = self.print_annotation(annotated_type)
 
-            if kind.is_named() and not any(arg.startswith("*") for arg in args):
-                args.append("*")
+            if kind.is_named() and not any(arg.name.startswith("*") for arg in args):
+                args.append(ArgSig("*"))
 
             if arg_.initializer:
-                if not annotation:
+                if not typename:
                     typename = self.get_str_type_of_node(arg_.initializer, True, False)
-                    if typename == "":
-                        annotation = "=..."
-                    else:
-                        annotation = f": {typename} = ..."
-                else:
-                    annotation += " = ..."
-                arg = name + annotation
             elif kind == ARG_STAR:
-                arg = f"*{name}{annotation}"
+                name = f"*{name}"
             elif kind == ARG_STAR2:
-                arg = f"**{name}{annotation}"
-            else:
-                arg = name + annotation
-            args.append(arg)
-        retname = None
+                name = f"**{name}"
+
+            args.append(ArgSig(name, typename, default=bool(arg_.initializer)))
+        return args
+
+    def _get_func_return(self, o: FuncDef, ctx: FunctionContext) -> str | None:
         if o.name != "__init__" and isinstance(o.unanalyzed_type, CallableType):
             if isinstance(get_proper_type(o.unanalyzed_type.ret_type), AnyType):
                 # Luckily, a return type explicitly annotated with "Any" has
                 # type "UnboundType" and will enter the else branch.
-                retname = None  # implicit Any
+                return None  # implicit Any
             else:
-                retname = self.print_annotation(o.unanalyzed_type.ret_type)
-        elif o.abstract_status == IS_ABSTRACT or o.name in METHODS_WITH_RETURN_VALUE:
+                return self.print_annotation(o.unanalyzed_type.ret_type)
+        if o.abstract_status == IS_ABSTRACT or o.name in METHODS_WITH_RETURN_VALUE:
             # Always assume abstract methods return Any unless explicitly annotated. Also
             # some dunder methods should not have a None return type.
-            retname = None  # implicit Any
-        elif o.name in KNOWN_MAGIC_METHODS_RETURN_TYPES:
-            retname = KNOWN_MAGIC_METHODS_RETURN_TYPES[o.name]
-        elif has_yield_expression(o) or has_yield_from_expression(o):
+            return None  # implicit Any
+        retname = infer_method_ret_type(o.name)
```

Comment:
`KNOWN_MAGIC_METHODS_RETURN_TYPES` is replaced by `infer_method_ret_type`, which is now used by both types of generators.